api/src/wfl/executor.clj

```diff
@@ -428,6 +446,91 @@
      executor
      (jdbc/query tx [(format query details) status]))))
 
+(defn ^:private reference-to-workflow-records
+  "Find all submission IDs associated with `workflows`.
+  Return a mapping of the submission reference (ex. snapshot reference id)
+  to their associated workflow records."
+  [tx {:keys [details] :as executor} workflows]
+  (throw-if-no-details-table tx executor)
+  (let [workflow-ids (util/to-quoted-comma-separated-list (map :uuid workflows))
+        _            (-> "%s Fetching submission IDs linked to workflows %s"
+                         (format (log-prefix executor) workflow-ids)
+                         log/debug)
+        ;; Presently Rawls doesn't support submitting a subset of a snapshot:
+        ;; If we wish to retry 1+ workflows stemming from a snapshot,
+        ;; we must resubmit the snapshot in its entirety.
+        query "SELECT * FROM %s
+               WHERE submission IN
+                 (SELECT DISTINCT submission FROM %s
+                 WHERE workflow IN %s)"]
+    (->> (format query details details workflow-ids)
+         (jdbc/query tx)
+         (group-by :reference))))
+
+(defn ^:private entity-from-id
+  "Use `fromSource` to direct fetch of existing `workspace` entity
+  identified by `entity-id`."
```

Comment:
This doesn't make sense to me, at least `fromSource` is being used in a way that I didn't intend. `fromSource` takes whatever the `Source` emitted and converts it into something that the executor can understand (i.e. an Entity`). Ideally, one would pattern match on the entity to tell what kind of thing your dealing with. This being clojure, nearest thing we can do is use tag-value pairs:

```clojure
(when-let [obj (peek-source s)]
  (let [[tag value :as entity] (coerce fromSource obj)]
    (case tag
      "Reference" ;; use reference
      "Row"       ;; use row
``` 

Thereafter, the executor knows what that `entity` is and `fromSource` is meaningless - one just needs to a case analysis on the entity to figure out what to do with it.

Sadly this isn't how it was written because we were in a rush. I maintain that using `fromSource` for any other purpose is the wrong thing to do.

Reply:
File a ticket to switch `reference` in executor details table to `entity` that's like `[tag value]` pair, serialized/deserialized as needed.

Reply:
For now we accept that we only support references.  Ticket to support generic entity submission: https://broadinstitute.atlassian.net/browse/GH-1422